app/push_notification.go

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"hash/fnv"
+	"net/http"
+	"strings"
+
+	l4g "github.com/alecthomas/log4go"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/utils"
+	"github.com/nicksnyder/go-i18n/i18n"
+)
+
+const PUSH_NOTIFICATION_HUB_WORKERS = 1000
+const PUSH_NOTIFICATIONS_HUB_BUFFER_PER_WORKER = 50
+
+type PushNotificationsHub struct {
+	Channels []chan PushNotification
+}
+
+type PushNotification struct {
+	notificationType string
+	userId           string
+	channelId        string
+	post             *model.Post
+	user             *model.User
+	channel          *model.Channel
+	senderName       string
+	channelName      string
+	wasMentioned     bool
+}
+
+func (a *App) sendPushNotificationSync(post *model.Post, user *model.User, channel *model.Channel, senderName, channelName string, wasMentioned bool) *model.AppError {
+	sessions, err := a.getMobileAppSessions(user.Id)
+	if err != nil {
+		return err
+	}
+
+	if channel.Type == model.CHANNEL_DIRECT {
+		channelName = senderName
+	}
+
+	msg := model.PushNotification{}
+	if badge := <-a.Srv.Store.User().GetUnreadCountFromMaster(user.Id); badge.Err != nil {
+		msg.Badge = 1
+		l4g.Error("We could not get the unread message count for the user", user.Id, badge.Err)
+	} else {
+		msg.Badge = int(badge.Data.(int64))
+	}
+
+	msg.Type = model.PUSH_TYPE_MESSAGE
+	msg.TeamId = channel.TeamId
+	msg.ChannelId = channel.Id
+	msg.PostId = post.Id
+	msg.RootId = post.RootId
+	msg.ChannelName = channel.Name
+	msg.SenderId = post.UserId
+
+	if ou, ok := post.Props["override_username"].(string); ok {
+		msg.OverrideUsername = ou
+	}
+
+	if oi, ok := post.Props["override_icon_url"].(string); ok {
+		msg.OverrideIconUrl = oi
+	}
+
+	if fw, ok := post.Props["from_webhook"].(string); ok {
+		msg.FromWebhook = fw
+	}
+
+	userLocale := utils.GetUserTranslations(user.Locale)
+	hasFiles := post.FileIds != nil && len(post.FileIds) > 0
+
+	msg.Message, msg.Category = a.getPushNotificationMessage(post.Message, wasMentioned, hasFiles, senderName, channelName, channel.Type, userLocale)
+
+	for _, session := range sessions {
+		tmpMessage := *model.PushNotificationFromJson(strings.NewReader(msg.ToJson()))
+		tmpMessage.SetDeviceIdAndPlatform(session.DeviceId)
+
+		l4g.Debug("Sending push notification to device %v for user %v with msg of '%v'", tmpMessage.DeviceId, user.Id, msg.Message)
+
+		a.Go(func(session *model.Session) func() {
+			return func() {
+				a.sendToPushProxy(tmpMessage, session)
+			}
+		}(session))
+
+		if a.Metrics != nil {
+			a.Metrics.IncrementPostSentPush()
+		}
+	}
+
+	return nil
+}
+
+func (a *App) getChannelFromUserId(userId string) chan PushNotification {
+	h := fnv.New32a()
+	h.Write([]byte(userId))
+	chanIdx := h.Sum32() % PUSH_NOTIFICATION_HUB_WORKERS
+	return a.PushNotificationsHub.Channels[chanIdx]
+}
```

Comment:
I get the channel based on the `userId` which ensure that the push notifications and push clear notifications actions are executed in order for each user (to avoid possible race conditions).